internal/profiling/profiling.go

```diff
@@ -0,0 +1,251 @@
+	GoID int64
+}
+
+// NewTimer creates and returns a new Timer object. This is useful when you
+// don't already have a Stat object to associate this Timer with; for example,
+// before the context of a new RPC query is created, a Timer may be needed to
+// measure transport-related operations.
+//
+// Use AppendTimer to append the returned Timer to a Stat.
+func NewTimer(timerTag string) *Timer {
+	return &Timer{
+		TimerTag: timerTag,
+		Begin:    time.Now(),
+		GoID:     goid(),
+	}
+}
+
+// A Stat is a collection of Timers that represent timing information for
+// different components within this Stat. For example, a Stat may be used to
+// reference the entire lifetime of an RPC request, with Timers within it
+// representing different components such as encoding, compression, and
+// transport.
+//
+// The user is expected to use the included helper functions to do operations
+// on the Stat such as creating or appending a new timer. Direct operations on
+// the Stat's exported fields (which are exported for encoding reasons) may
+// lead to data races.
+type Stat struct {
+	// Each Stat has a tag associated with it to identify it in post-processing.
+	// This is can be any arbitrary string.
+	StatTag string
+	// Stats may also need to store other unstructured information specific to
+	// this stat. For example, a StreamStat will use these bytes to encode the
+	// connection number and the stream ID for each RPC, thereby uniquely
+	// identifying it. The underlying encoding for this is unspecified.
+	Metadata []byte
+	// A collection of Timers and a mutex for append operations on the slice.
+	mu     sync.Mutex
+	Timers []Timer
+}
+
+// A power of two that's large enough to hold all timers within an average RPC
+// request (defined to be a unary request) without any reallocation. A typical
+// unary RPC creates 80-100 timers for various things. While this number is
+// purely anecdotal and may change in the future as the resolution of profiling
+// increases or decreases, it serves as a good estimate for what the initial
+// allocation size should be.
+const defaultStatAllocatedTimers int32 = 128
+
+// NewStat creates and returns a new Stat object.
+func NewStat(statTag string) *Stat {
+	return &Stat{
+		StatTag: statTag,
+		Timers:  make([]Timer, 0, defaultStatAllocatedTimers),
+	}
+}
+
+// NewTimer creates a Timer object within the given stat if stat is non-nil.
+// The value passed in timerTag will be attached to the newly created Timer.
+// NewTimer also automatically sets the Begin value of the Timer to the current
+// time.
+//
+// The user is expected to call stat.Egress with the returned index as argument
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
```

Comment:
`NewTimer` isn't allocation free, and does allocate a `Timer`, right?

Did you mean to `make([]Timer, defaultStatAllocatedTimers)` in `NewStat`, and set the fields of a timer in `NewTimer`? (This makes `defaultStatAllocatedTimers` a hard limit of the slice length, though)

Or this was the old behavior, you changed the implementation, but forgot to update the comment?

Reply:
Yeah, I previous used to do `Timers[i].field = value` with a fixed size Timer array.

But I did some benchmarks and I was surprised to find that doing `append(slice, &Timer{})` is faster than `append(slice, Timer{})` -- I wrongly assumed that `append(slice, Timer{})` would be faster. Changed the API slightly to use a `*Timer`.